src/net_processing.cpp

```diff
@@ -2092,6 +2094,21 @@ void PeerManagerImpl::StartScheduledTasks(CScheduler& scheduler)
     scheduler.scheduleFromNow([&] { ReattemptInitialBroadcast(scheduler); }, delta);
 }
 
+void PeerManagerImpl::ActiveTipChange(const CBlockIndex* pindexNew, bool is_ibd)
+{
+    AssertLockNotHeld(m_mempool.cs);
```

Comment:
I think you meant `m_tx_download_mutex`

Reply:
No, I think the `LOCK(m_tx_download_mutex)` already checks that it isn't already held, since it's a `Mutex`.

I guess this seems a bit random, but I'm trying to establish the lock hierarchy as `m_tx_download_mutex`, then `m_mempool.cs` (which is a `RecursiveMutex`). That's also why we can only fire `ActiveTipChanged` after releasing the mempool mutex. Later code added to `TxDownloadManager` will call `CTxMemPool` functions that take the mempool lock (see e.g. https://github.com/bitcoin/bitcoin/pull/30110/files#diff-5d7e42fb1e904f54956c0975a9307c58d8d36d77ad9da8bcd8989437356e3229R264 or grep `exists` within txdownload_impl).

Reply:
Ah I was wrong about `LOCK` doing the assert, added that as well.

Reply:
As far as I understand, `AssertLockNotHeld(m_mempool.cs);` is an attempt to force the mutex hierarchy mentioned in here: https://github.com/bitcoin/bitcoin/blob/c85accecafc20f6a6ae94bdf6cdd3ba9747218fd/src/net_processing.cpp#L784

As `m_mempool.cs` is not used in this function, I suggest to add a comment to prevent tempting of removing this line on the "clean up" purpose in the future.